services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+   * 3. run a preferred replica election
+   * </pre>
+   */
+  void rebalanceMonitoredTopic(TopicState topicState) {
+    try {
+      LOG.debug(_serviceName + ": broker count " + topicState._allBrokers.size() + " partitionCount " + topicState._partitionInfo.size() + ".");
+
+      if (topicState.partitionsLow(_expectedPartitionBrokerRatio)) {
+        int idealPartitionCount = _partitionsPerBroker * topicState._allBrokers.size();
+        int addPartitionCount = idealPartitionCount - topicState._partitionInfo.size();
+        LOG.info(_serviceName + ": adding " + addPartitionCount + " partitions.");
+        addPartitions(_zkUtils, idealPartitionCount);
+        waitForAddedPartitionsToBecomeActive(idealPartitionCount);
+        topicState = topicState();
+      }
+      if (topicState.brokerMissingPartition()) {
+        LOG.info(_serviceName + ": rebalancing monitored topic.");
+        waitForOtherAssignmentsToComplete();
+        reassignPartitions(topicState._allBrokers, topicState._partitionInfo.size());
+        waitForPartitionReassignmentToComplete();
+        topicState = topicState();
+      }
+      if (topicState.brokerNotElectedLeader()) {
+        LOG.info(_serviceName + ": running preferred replica election.");
+        runPreferredElection(_zkUtils, topicState._partitionInfo);
+      }
+    } catch (InterruptedException ie) {
+      throw new IllegalStateException(ie);
+    }
+  }
+
+  /**
+   *  Polling...
+   */
+  private void waitForAddedPartitionsToBecomeActive(int expectedNumberOfPartitions) {
+    long timeout = System.currentTimeMillis() + 1000 * 60;
+
+    //Using ZkUtils instead of the consumer because the consumer seems to cache the last answer it got which was the old
+    //number of partitions.
+    scala.collection.mutable.ArrayBuffer<String> scalaTopic = new scala.collection.mutable.ArrayBuffer<>();
+    scalaTopic.$plus$eq(_topic);
+    while (System.currentTimeMillis() < timeout) {
+      if (_zkUtils.getPartitionAssignmentForTopics(scalaTopic).apply(_topic).size()  == expectedNumberOfPartitions) {
+        return;
+      }
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+    throw new IllegalStateException("Waiting for additional partitions to appear timed out.");
+  }
+
+  /**
+   * If there is some other assignment going then wait for it to complete.
+   */
+  private void waitForOtherAssignmentsToComplete() throws InterruptException {
+    while (!_zkUtils.getPartitionsBeingReassigned().isEmpty()) {
+      try {
+        LOG.debug("Waiting for current partition assignment to be complete.");
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  private void waitForPartitionReassignmentToComplete() throws InterruptedException {
```

Comment:
It is probably good enough to simply check if there is ongoing partition reassignment without looking at the topic name. The reason is that we only begin topic reassignment if there is no ongoing topic reassignment. In the rare case there someone else initiated a topic reassignment after our check but before we create our reassignment, it should be OK to wait for that topic as well.

Thus we can probably replace this method with waitForOtherAssignmentsToComplete(). Maybe rename it `waitForOngoingAssignmentsToComplete`